targets/TARGET_ARM_FM/TARGET_FVP_MPS2/serial_api.c

```diff
@@ -0,0 +1,384 @@
+{
+    uart_irq(CMSDK_UART3->INTSTATUS & 0x3, 2, (CMSDK_UART_TypeDef *)CMSDK_UART3);
+}
+void uart3_irq()
+{
+    uart_irq(CMSDK_UART4->INTSTATUS & 0x3, 3, (CMSDK_UART_TypeDef *)CMSDK_UART4);
+}
+
+void serial_irq_handler(serial_t *obj, uart_irq_handler handler, uint32_t id)
+{
+    irq_handler = handler;
+    uart_data[obj->index].serial_irq_id = id;
+}
+
+static void serial_irq_set_internal(serial_t *obj, SerialIrq irq, uint32_t enable)
+{
+    /* Declare a variable of type IRQn, initialise to 0  */
+    IRQn_Type irq_n = (IRQn_Type)0;
+    uint32_t vector = 0;
+    switch ((int)obj->uart) {
+        case UART_0: {
+                irq_n   = ((irq == TxIrq) ?  UARTTX0_IRQn : UARTRX0_IRQn);
+                vector  = (uint32_t)&uart0_irq;
+            }
+            break;
+
+        case UART_1: {
+                irq_n   = ((irq == TxIrq) ?  UARTTX1_IRQn : UARTRX1_IRQn);
+                vector  = (uint32_t)&uart1_irq;
+            }
+            break;
+        case UART_2: {
+                irq_n   = ((irq == TxIrq) ?  UARTTX3_IRQn : UARTRX3_IRQn);
+                vector  = (uint32_t)&uart2_irq;
+            }
+            break;
+        case UART_3: {
+                irq_n   = ((irq == TxIrq) ?  UARTTX4_IRQn : UARTRX4_IRQn);
+                vector  = (uint32_t)&uart3_irq;
+            }
+            break;
+    }
+
+    if (enable) {
+        if (irq == TxIrq) {
+            /* Transmit IRQ, set appripriate enable */
+
+            /* set TX interrupt enable in CTRL REG */
+            obj->uart->CTRL |= CMSDK_UART_CTRL_TXIRQEN_Msk;
+        } else {
+            /* set Rx interrupt on in CTRL REG */
+            obj->uart->CTRL |= CMSDK_UART_CTRL_RXIRQEN_Msk;
+        }
+        NVIC_SetVector(irq_n, vector);
+        NVIC_EnableIRQ(irq_n);
+
+    } else {
+        /*      Disable   IRQ   */
+
+        obj->uart->CTRL &= ~(1 << (irq + 2));
+
+        NVIC_DisableIRQ(irq_n);
+    }
+}
+
+void serial_irq_set(serial_t *obj, SerialIrq irq, uint32_t enable)
+{
+    serial_irq_set_internal(obj, irq, enable);
+}
+
+/******************************************************************************
+ * READ/WRITE
+ ******************************************************************************/
+int serial_getc(serial_t *obj)
+{
+    while (serial_readable(obj) == 0);
+    int data = obj->uart->DATA;
+    return data;
+}
+
+void serial_putc(serial_t *obj, int c)
+{
+    while (serial_writable(obj) == 0);
+    obj->uart->DATA = c;
+}
+
+int serial_readable(serial_t *obj)
+{
+    return obj->uart->STATE & 0x2;
+}
+
+int serial_writable(serial_t *obj)
+{
+    return !(obj->uart->STATE & 0x1);
+}
+
+void serial_clear(serial_t *obj)
+{
+    obj->uart->DATA = 0x00;
+}
+
+void serial_pinout_tx(PinName tx)
+{
+    pinmap_pinout(tx, PinMap_UART_TX);
+}
+
+void serial_break_set(serial_t *obj)
+{
+}
+
+void serial_break_clear(serial_t *obj)
+{
+}
+void serial_set_flow_control(serial_t *obj, FlowControl type, PinName rxflow, PinName txflow)
+{
```

Comment:
why flow control is emtpy? is it enabled? will it be supported?

Similar is to pinmap functionality (assume that target does not have any pinmap functionality, neither break set/clear) thus they are emptY?

Reply:
the fastmodels and the hardware it modeling after don't support for flow control.
The details can be found in bellow reference guide:
*  Ref. [DDI0479C_cortex_m_system_design_kit_r1p0_trm.pdf ](http://infocenter.arm.com/help/topic/com.arm.doc.ddi0479c/DDI0479C_cortex_m_system_design_kit_r1p0_trm.pdf)